# Перевірка статиcтичних гіпотез. Z-test.



0. Зчитайте дані з `data.csv` у змінну data, яка має тип pandas.DataFrame. Ми будемо далі працювати з цією змінною.

In [5]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats

# Завантаження даних
data = pd.read_csv('D:/DATAS/Python/data/Модуль 15/data.csv')

Запустіть код нижче. Для коректної робити всіх подальших методів дані мають бути у вигляді одновимірного `numpy.ndarray` вектора та мати розмірність (100,). Така розмірність означає, що дані - одновимірні (колонка), якщо б розмірність була (100, 1), то дані сприймаються методами як двовимірні (таблиця), хоча для нас виглядати можуть так само.

In [6]:
data = data.data.values
data.shape

(100,)

**Завдання 1**. Зчитані дані - це сума покупок в доларах наших покупців на сайті протягом липня 2025 року.

До цього місяця, ми мали продажі в середньому на чек $20.

Необхідно зʼясувати, чи вийшло у нас статистично значущо **підвищити** середній чек за липень 2025?

Аби дати відповідь - ми проведемо z-test на рівні значущості $\alpha=0.05$ двома способами. В цьому завданні ми будемо виконувати обчислення "вручну" з використанням бібліотек numpy та scipy.stats подібно до прикладу в лекції. Для цього виконайте наступні 7 кроків. Правильне виконання кожного з кроків оцінюється в 1 бал.

1. Запишіть параметри задачі у змінні Python:
    - `sample_mean` - $\bar{x}$, середнє значення в вибірці
    - `population_mean` - $\mu_0$, середнє значення в популяції (тобто те, з яким ми порівнюємо середнє вибіркове значення)
    - `population_std` - $\sigma$, вибіркове стандартне відхилення, яке ми вважаємо, що є рівним ст. відх. популяції, адже маємо достатньо велику вибірку.
    - `sample_size` - $n$, розмір вибірки
    - `alpha` - $\alpha$ рівень значущості
    
    **Увага!** Для обчислення стандартного відхилення маємо скористатись функцією `np.std(your_dataframe, ddof=1)`. Чому так - розбираємо в лекції про t-test.


2. Визначте для цієї задачі:
    - якою є гіпотеза $H_0$
    - якою є альтернативна гіпотеза $H_a$
    - з яким типом тесту ми маємо справу - лівосторонній, правосторонній чи двосторонній.

3. Обчисліть стандартну помилку SE.
4. Розрахуйте z-статистику.
5. Знайдіть критичне z-значення з допомогою бібліотеки stats.
6. Обчисліть p-value з допомогою бібліотеки stats.
7. Прийміть рішення, чи відхиляєте ви гіпотезу $H_0$. Для прийняття рішення зробіть і порівняння z-статистики з критичним значенням, і проаналізуйте p-value.

**Розв'язок:**

1. Записую параметри у змінні:

In [13]:
sample_mean = np.mean(data)                      # середнє значення у вибірці
population_mean = 20                             # середнє значення у популяції (історичне за умовою задачі)      
population_std = np.std(data, ddof=1)            # вибіркове стандартне відхилення
sample_size = len(data)                          # розмір вибірки
alpha = 0.05                                     # рівень значущості (за умовою задачі)

2. Формулюю гіпотези та визначаю вид тесту:
   - Нульова гіпотеза $Hо$ : середній чек у липні 2025 року залишився як в минулих місяцях і складає 20 доларів - $\mu_о$ = 20.
   - Альтернативна гіпотеза $Ha$ : середній чек у липні збільшився - $\mu_а$ > 20.
   - Це правосторонній тест.

3. Обчислюю стандартну помилку SE:

In [19]:
standard_error = population_std / np.sqrt(sample_size)

print("Формула: SE = σ / √n")
print()
print(f"SE = {population_std:.2f} / √{sample_size}")
print(f"SE = {population_std:.2f} / {np.sqrt(sample_size):.0f}")
print(f"SE = {standard_error:.3f}")
print()
print("Висновок: В середньому, вибіркове середнє відхиляється")
print(f"від справжнього середнього на ±{standard_error:.3f} доларів")

Формула: SE = σ / √n

SE = 6.25 / √100
SE = 6.25 / 10
SE = 0.625

Висновок: В середньому, вибіркове середнє відхиляється
від справжнього середнього на ±0.625 доларів


4. Розраховую z-статистику:

In [45]:
z_statistic = (sample_mean - population_mean) / standard_error

print("Формула: z = (x̄ - μ₀) / SE")
print()
print(f"z = ({sample_mean:.2f} - {population_mean}) / {standard_error:.3f}")
print(f"z = {sample_mean - population_mean:.2f} / {standard_error:.3f}")
print(f"z = {z_statistic:.3f}")
print()
print("Висновок:")
print(f"Вибіркове середнє відхиляється на {z_statistic:.3f} стандартних помилок")
print("від гіпотетичного середнього популяції.")
print()
if z_statistic > 3:
    print("Це дуже велике відхилення (більше 3 стандартних помилок).")
elif z_statistic > 2:
    print("Це велике відхилення (між 2 та 3 стандартними помилками).")
elif z_statistic > 1:
    print("Це помірне відхилення (між 1 та 2 стандартними помилками).")
else:
    print("Це невелике відхилення (менше 1 стандартної помилки).")

Формула: z = (x̄ - μ₀) / SE

z = (19.38 - 20) / 0.625
z = -0.62 / 0.625
z = -0.995

Висновок:
Вибіркове середнє відхиляється на -0.995 стандартних помилок
від гіпотетичного середнього популяції.

Це невелике відхилення (менше 1 стандартної помилки).


5. Шукаю критичне z-значення:

In [44]:
# Для правостороннього тесту з α = 0.05
z_critical = stats.norm.ppf(1 - alpha)

print(f"Для рівня значущості α = {alpha} (для правостороннього тесту):")
print(f"критичне z-значення з таблиці = {z_critical:.3f}")
print()
print(f"Якщо z-статистика > {z_critical:.3f}, то результат настільки незвичайний,")
print("що ми не віримо у випадковість за гіпотезою Но.")
print(f"Ймовірність отримати z > {z_critical:.3f} випадково = {alpha*100}%")

Для рівня значущості α = 0.05 (для правостороннього тесту):
критичне z-значення з таблиці = 1.645

Якщо z-статистика > 1.645, то результат настільки незвичайний,
що ми не віримо у випадковість за гіпотезою Но.
Ймовірність отримати z > 1.645 випадково = 5.0%


6. Обчислюю p-value:

In [43]:
# Для правостороннього тесту: p-value = P(Z > z_statistic)
p_value = 1 - stats.norm.cdf(z_statistic)

print(f"p-value = {p_value:.2f}")
print()
print(f"Якщо гіпотеза H₀ правдива (середній чек у липні 2025 року так і залишається 20 доларів), ")
print(f"то ймовірність випадково отримати середнє значення чеку ≥ {sample_mean:.2f} становить {p_value:.2f} або приблизно {p_value*100:.2f}%.")
print()
if p_value < 0.001:
    print("Це НАДЗВИЧАЙНО малоймовірно! (менше 0.1%).")
elif p_value < 0.01:
    print("Це дуже малоймовірно! (менше 1%).")
elif p_value < 0.05:
    print("Це малоймовірно! (менше 5%).")
else:
    print("Це цілком можливо випадково.")

p-value = 0.84

Якщо гіпотеза H₀ правдива (середній чек у липні 2025 року так і залишається 20 доларів), 
то ймовірність випадково отримати середнє значення чеку ≥ 19.38 становить 0.84 або приблизно 84.02%.

Це цілком можливо випадково.


7. Приймаю рішення, чи відхиляю гіпотезу $H_о$:
- Метод 1: Порівняння z-статистики з критичним значенням:

In [39]:
print(f"z-статистика = {z_statistic:.3f}")
print(f"Критичне значення = {z_critical:.3f}")
print(f"Порівняння: {z_statistic:.3f} {'>' if z_statistic > z_critical else '≤'} {z_critical:.3f}")

if z_statistic > z_critical:
    print("Висновок: z-статистика ПЕРЕВИЩУЄ критичне значення.")
    decision1 = "ВІДХИЛЯЄМО H₀."
else:
    print("Висновок: z-статистика НЕ перевищує критичне значення.")
    decision1 = "НЕ ВІДХИЛЯЄМО H₀."

print(f"Рішення: {decision1}")

z-статистика = -0.995
Критичне значення = 1.645
Порівняння: -0.995 ≤ 1.645
Висновок: z-статистика НЕ перевищує критичне значення.
Рішення: НЕ ВІДХИЛЯЄМО H₀.


 - Метод 2: Порівняння p-value з рівнем значущості (p-value дає точну ймовірність):

In [41]:
print(f"p-value = {p_value:.2f}")
print(f"Рівень значущості α = {alpha}")
print(f"Порівняння: {p_value:.2f} {'<' if p_value < alpha else '≥'} {alpha}")

if p_value < alpha:
    print("Висновок: p-value МЕНШЕ за рівень значущості.")
    decision2 = "ВІДХИЛЯЄМО H₀."
else:
    print("Висновок: p-value НЕ менше за рівень значущості.")
    decision2 = "НЕ ВІДХИЛЯЄМО H₀."

print(f"Рішення: {decision2}")

p-value = 0.84
Рівень значущості α = 0.05
Порівняння: 0.84 ≥ 0.05
Висновок: p-value НЕ менше за рівень значущості.
Рішення: НЕ ВІДХИЛЯЄМО H₀.


**Висновок:**

Завдяки двом незалежним способам перевірки, через z‑статистику та p‑value, підтвердилось одне й те саме:
- z‑статистика = -0.995 показує, що вибіркове середнє навіть трохи менше за історичне, і відхилення менше за критичне значення z_critical=1.645.
- p‑value = 0.84 значно перевищує рівень значущості $\alpha$ =0.05, тобто ймовірність отримати такі дані при $H₀$ дуже велика.

**Не відхиляємо $H₀$.** Це означає, що статистично значущого підвищення середнього чеку у липні 2025 року не відбулося, тому середній чек залишився на рівні близько 20 доларів, як і в попередні місяці 2025 року.


**Завдання 2.** Виконайте обчислення z-test з використанням бібліотеки statsmodels.

Отримайте z-статистику та р-значення.

Виведіть p-значення та зробіть висновок, чи ми маємо достатньо доказів, аби стверджувати, що середній чек зріс.

Чи зійшлись значення z-статистику та р-значення в цьому завданні з попередніми обчисленнями?

Завдання оцінюється в 3 бали.

In [8]:
from statsmodels.stats import weightstats as stests

In [51]:
# Виконуємо правосторонній z-тест
z_stat_sm, p_val_sm = stests.ztest(data,                   
                                   value=population_mean,  # μ₀ = 20
                                   alternative='larger')   # H₁: μа > 20

print(f"Результати z-тесту:")
print(f"  Z-статистика: {z_stat_sm:.3f}")
print(f"  P-value: {p_val_sm:.2f}")

# Інтерпретація результатів
print("Висновок:")
if p_val_sm < alpha:
    print(f"  ВІДХИЛЯЄМО H₀.")
    print(f"  p-value ({p_val_sm:.2f}) < α ({alpha})")
    print("  Середній чек у липні 2025 року виріс порівняно з іншими місяцями.")
else:
    print(f"  НЕ ВІДХИЛЯЄМО H₀.")
    print(f"  p-value ({p_val_sm:.2f}) ≥ α ({alpha})")
    print("  Статистично значущого підвищення середнього чеку у липні 2025 року не відбулося.")
print()
print("Порівняння з ручними розрахунками:")
print(f"Z-статистика:")
print(f"  Ручний розрахунок: {z_statistic:.3f}")
print(f"  Statsmodels:       {z_stat_sm:.3f}")
print(f"  Різниця:           {abs(z_statistic - z_stat_sm):.3f}")
print("Висновок: різниці між ручними розрахунками та обчисленнями з використанням бібліотеки statsmodels немає.")

Результати z-тесту:
  Z-статистика: -0.995
  P-value: 0.84
Висновок:
  НЕ ВІДХИЛЯЄМО H₀.
  p-value (0.84) ≥ α (0.05)
  Статистично значущого підвищення середнього чеку у липні 2025 року не відбулося.

Порівняння з ручними розрахунками:
Z-статистика:
  Ручний розрахунок: -0.995
  Statsmodels:       -0.995
  Різниця:           0.000
Висновок: різниці між ручними розрахунками та обчисленнями з використанням бібліотеки statsmodels немає.
